contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+        require(referrer != address(0), "!referrer");
+        require(fee < MAX_PERCENTAGE, "Fee exceeds maximum");
+
+        referralFees[referrer] = fee;
+    }
+
+    /************************************************
+     *  OFFER CREATION AND SETTLEMENT
+     ***********************************************/
+
+    /**
+     * @notice Create a new offer available for swap
+     * @param offeredToken token offered by seller
+     * @param biddingToken token asked by seller
+     * @param minPrice minimum price of offeredToken in terms of biddingToken
+     * @param minBidSize minimum size allowed in terms of biddingToken
+     * @param totalSize amount of offeredToken offered by seller
+     */
+    function createOffer(
+        address offeredToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(offeredToken != address(0), "!offeredToken");
+        require(biddingToken != address(0), "!biddingToken");
+        require(minPrice > 0, "!minPrice");
+        require(minBidSize > 0, "!minBidSize");
+        require(totalSize > 0, "!totalSize");
+
+        offersCounter += 1;
+
+        swapId = offersCounter;
+
+        swapOffers[swapId] = Offer({
+            seller: msg.sender,
+            isOpen: true,
+            offeredToken: offeredToken,
+            biddingToken: biddingToken,
+            offeredTokenDecimals: IERC20Detailed(offeredToken).decimals(),
+            minBidSize: minBidSize,
+            minPrice: minPrice,
+            totalSize: totalSize,
+            availableSize: totalSize,
+            highestPrice: 0,
+            totalSales: 0
+        });
+
+        emit NewOffer(
+            swapId,
+            msg.sender,
+            offeredToken,
+            biddingToken,
+            minPrice,
+            minBidSize,
+            totalSize
+        );
+    }
+
+    /**
+     * @notice Settles the swap offering by iterating through the bids
+     * @param swapId unique identifier of the swap offer
+     * @param bids bids for swaps
+     */
+    function settleOffer(uint256 swapId, Bid[] calldata bids)
+        external
+        override
+    {
+        Offer storage offer = swapOffers[swapId];
+
+        require(offer.seller != address(0), "Offer does not exist");
+        require(msg.sender == offer.seller, "Only seller can settle");
+        require(offer.isOpen, "Offer already closed");
+
+        for (uint256 i = 0; i < bids.length; i++) {
+            _swap(offer, bids[i]);
+        }
+
+        if (offer.availableSize == 0) {
+            offer.isOpen = false;
+
+            emit SettleOffer(swapId);
+        }
+    }
+
+    /**
+     * @notice Close offer
+     * @param swapId swapId unique identifier of the swap offer
+     */
+    function closeOffer(uint256 swapId) external override {
+        Offer storage offer = swapOffers[swapId];
```

Comment:
closeOffer can be used to `delete` the entire Offering struct so we can [get gas refunds](https://ethereum.stackexchange.com/questions/594/how-do-gas-refunds-work)

Reply:
Was thinking the `closeOffer` is for our ThetaVault to prematurely close the offer when `availableSize != 0`.